File: opendc-web/opendc-web-server/src/main/java/org/opendc/web/server/service/ScenarioService.java
/*
 * Copyright (c) 2023 AtLarge Research
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy
 * of this software and associated documentation files (the "Software"), to deal
 * in the Software without restriction, including without limitation the rights
 * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
 * copies of the Software, and to permit persons to whom the Software is
 * furnished to do so, subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included in all
 * copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
 * SOFTWARE.
 */

package org.opendc.web.server.service;

import java.time.Instant;
import java.util.List;
import javax.enterprise.context.ApplicationScoped;
import org.opendc.web.proto.JobState;
import org.opendc.web.server.model.Job;
import org.opendc.web.server.model.Portfolio;
import org.opendc.web.server.model.ProjectAuthorization;
import org.opendc.web.server.model.Scenario;
import org.opendc.web.server.model.Topology;
import org.opendc.web.server.model.Trace;
import org.opendc.web.server.model.Workload;

/**
 * Service for managing {@link Scenario}s.
 */
@ApplicationScoped
public final class ScenarioService {
    /**
     * The service for managing the user accounting.
     */
    private final UserAccountingService accountingService;

    /**
     * Construct a {@link ScenarioService} instance.
     *
     * @param accountingService The {@link UserAccountingService} instance to use.
     */
    public ScenarioService(UserAccountingService accountingService) {
        this.accountingService = accountingService;
    }

    /**
     * List all {@link Scenario}s that belong a certain portfolio.
     */
    public List<org.opendc.web.proto.user.Scenario> findAll(String userId, long projectId, int number) {
        // User must have access to project
        ProjectAuthorization auth = ProjectAuthorization.findByUser(userId, projectId);

        if (auth == null) {
            return List.of();
        }

        return Scenario.findByPortfolio(projectId, number).stream()
                .map((s) -> toUserDto(s, auth))
                .toList();
    }

    /**
     * Obtain a {@link Scenario} by identifier.
     */
    public org.opendc.web.proto.user.Scenario findOne(String userId, long projectId, int number) {
        // User must have access to project
        ProjectAuthorization auth = ProjectAuthorization.findByUser(userId, projectId);

        if (auth == null) {
            return null;
        }

        Scenario scenario = Scenario.findByProject(projectId, number);

        if (scenario == null) {
            return null;
        }

        return toUserDto(scenario, auth);
    }

    /**
     * Delete the specified scenario.
     */
    public org.opendc.web.proto.user.Scenario delete(String userId, long projectId, int number) {
        // User must have access to project
        ProjectAuthorization auth = ProjectAuthorization.findByUser(userId, projectId);

        if (auth == null) {
            return null;
        } else if (!auth.canEdit()) {
            throw new IllegalStateException("Not permitted to edit project");
        }

        Scenario entity = Scenario.findByProject(projectId, number);
        if (entity == null) {
            return null;
        }

        entity.delete();
        return toUserDto(entity, auth);
    }

    /**
     * Construct a new {@link Scenario} with the specified data.
     */
    public org.opendc.web.proto.user.Scenario create(
            String userId, long projectId, int portfolioNumber, org.opendc.web.proto.user.Scenario.Create request) {
        // User must have access to project
        ProjectAuthorization auth = ProjectAuthorization.findByUser(userId, projectId);

        if (auth == null) {
            return null;
        } else if (!auth.canEdit()) {
            throw new IllegalStateException("Not permitted to edit project");
        }

        Portfolio portfolio = Portfolio.findByProject(projectId, portfolioNumber);

        if (portfolio == null) {
            return null;
        }

        Topology topology = Topology.findByProject(projectId, (int) request.getTopology());
        if (topology == null) {
            throw new IllegalArgumentException("Referred topology does not exist");
        }

        Trace trace = Trace.findById(request.getWorkload().getTrace());
        if (trace == null) {
            throw new IllegalArgumentException("Referred trace does not exist");
        }

        var now = Instant.now();
        var project = auth.project;
        int number = project.allocateScenario(now);

        Scenario scenario = new Scenario(
                project,
                portfolio,
                number,
                request.getName(),
                new Workload(trace, request.getWorkload().getSamplingFraction()),
                topology,
                request.getPhenomena(),
                request.getSchedulerName());
        Job job = new Job(scenario, userId, now, portfolio.targets.getRepeats());

        // Fail the job if there is not enough budget for the simulation
        if (!accountingService.hasSimulationBudget(userId)) {
            job.state = JobState.FAILED;
        }

        scenario.job = job;
        portfolio.scenarios.add(scenario);
        scenario.persist();

        return toUserDto(scenario, auth);
    }

    /**
     * Convert a {@link Scenario} entity into a {@link org.opendc.web.proto.user.Scenario} DTO.
     */
    public static org.opendc.web.proto.user.Scenario toUserDto(Scenario scenario, ProjectAuthorization auth) {
        return new org.opendc.web.proto.user.Scenario(
                scenario.id,
                scenario.number,
                ProjectService.toUserDto(auth),
                PortfolioService.toSummaryDto(scenario.portfolio),
                scenario.name,
                toDto(scenario.workload),
                TopologyService.toSummaryDto(scenario.topology),
                scenario.phenomena,
                scenario.schedulerName,
                JobService.toUserDto(scenario.job));
    }

    /**
     * Convert a {@link Scenario} entity into a {@link org.opendc.web.proto.user.Scenario.Summary} DTO.
     */
    public static org.opendc.web.proto.user.Scenario.Summary toSummaryDto(Scenario scenario) {
        return new org.opendc.web.proto.user.Scenario.Summary(
                scenario.id,
                scenario.number,
                scenario.name,
                toDto(scenario.workload),
                TopologyService.toSummaryDto(scenario.topology),
                scenario.phenomena,
                scenario.schedulerName,
                JobService.toUserDto(scenario.job));
    }

    /**
     * Convert a {@link Scenario} into a runner-facing DTO.
     */
    public static org.opendc.web.proto.runner.Scenario toRunnerDto(Scenario scenario) {
        return new org.opendc.web.proto.runner.Scenario(
                scenario.id,
                scenario.number,
                PortfolioService.toRunnerDto(scenario.portfolio),
                scenario.name,
                toDto(scenario.workload),
                TopologyService.toRunnerDto(scenario.topology),
                scenario.phenomena,
                scenario.schedulerName);
    }

    /**
     * Convert a {@link Workload} entity into a DTO.
     */
    public static org.opendc.web.proto.Workload toDto(Workload workload) {
        return new org.opendc.web.proto.Workload(TraceService.toUserDto(workload.trace), workload.samplingFraction);
    }
}

File: opendc-web/opendc-web-api/src/main/kotlin/org/opendc/web/api/service/ScenarioService.kt
/*
 * Copyright (c) 2022 AtLarge Research
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy
 * of this software and associated documentation files (the "Software"), to deal
 * in the Software without restriction, including without limitation the rights
 * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
 * copies of the Software, and to permit persons to whom the Software is
 * furnished to do so, subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included in all
 * copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
 * SOFTWARE.
 */

package org.opendc.web.api.service

import org.opendc.web.api.model.*
import org.opendc.web.api.repository.*
import org.opendc.web.proto.user.Scenario
import java.time.Instant
import javax.enterprise.context.ApplicationScoped
import javax.inject.Inject

/**
 * Service for managing [Scenario]s.
 */
@ApplicationScoped
class ScenarioService @Inject constructor(
    private val projectRepository: ProjectRepository,
    private val portfolioRepository: PortfolioRepository,
    private val topologyRepository: TopologyRepository,
    private val traceRepository: TraceRepository,
    private val scenarioRepository: ScenarioRepository,
) {
    /**
     * List all [Scenario]s that belong a certain portfolio.
     */
    fun findAll(userId: String, projectId: Long, number: Int): List<Scenario> {
        // User must have access to project
        val auth = projectRepository.findOne(userId, projectId) ?: return emptyList()
        val project = auth.toUserDto()
        return scenarioRepository.findAll(projectId).map { it.toUserDto(project) }
    }

    /**
     * Obtain a [Scenario] by identifier.
     */
    fun findOne(userId: String, projectId: Long, number: Int): Scenario? {
        // User must have access to project
        val auth = projectRepository.findOne(userId, projectId) ?: return null
        val project = auth.toUserDto()
        return scenarioRepository.findOne(projectId, number)?.toUserDto(project)
    }

    /**
     * Delete the specified scenario.
     */
    fun delete(userId: String, projectId: Long, number: Int): Scenario? {
        // User must have access to project
        val auth = projectRepository.findOne(userId, projectId)

        if (auth == null) {
            return null
        } else if (!auth.role.canEdit) {
            throw IllegalStateException("Not permitted to edit project")
        }

        val entity = scenarioRepository.findOne(projectId, number) ?: return null
        val scenario = entity.toUserDto(auth.toUserDto())
        scenarioRepository.delete(entity)
        return scenario
    }

    /**
     * Construct a new [Scenario] with the specified data.
     */
    fun create(userId: String, projectId: Long, portfolioNumber: Int, request: Scenario.Create): Scenario? {
        // User must have access to project
        val auth = projectRepository.findOne(userId, projectId)

        if (auth == null) {
            return null
        } else if (!auth.role.canEdit) {
            throw IllegalStateException("Not permitted to edit project")
        }

        val portfolio = portfolioRepository.findOne(projectId, portfolioNumber) ?: return null
        val topology = requireNotNull(
            topologyRepository.findOne(
                projectId,
                request.topology.toInt()
            )
        ) { "Referred topology does not exist" }
        val trace =
            requireNotNull(traceRepository.findOne(request.workload.trace)) { "Referred trace does not exist" }

        val now = Instant.now()
        val project = auth.project
        val number = projectRepository.allocateScenario(auth.project, now)

        val scenario = Scenario(
            0,
            number,
            request.name,
            project,
            portfolio,
            Workload(trace, request.workload.samplingFraction),
            topology,
            request.phenomena,
            request.schedulerName
        )
        val job = Job(0, scenario, now, portfolio.targets.repeats)

        scenario.job = job
        portfolio.scenarios.add(scenario)
        scenarioRepository.save(scenario)

        return scenario.toUserDto(auth.toUserDto())
    }
}
